### DISCUSSION TITLE
Sigh, why weren't capsules kept for version 3, as it's a HUGE breaking change!?!

### DISCUSSION THREAD

**Discussion by kirkbushell** [area17/twill#1644] [Ideas]
I'm trying to migrate my twill cms installation to the latest, moving away from modules to capsules, and the amount of work to get this going (because there's no documentation, and some configuration doesn't seem to have any effect)... It's just another reason why I get so damn frustrated with this CMS.

I understand you guys are busy, I understand this is developed for your clients and so first used in-house, but you guys need to develop a better approach to releasing software.

Capsules were never a great idea, I put forth my arguments numerous times on this, and the amount of work required to enforce an opinionated structure is just ludicrous. The amount of configuration for this software is just... it boggles the damn mind.

This isn't the Laravel way. If anything, it's the antithesis of the Laravel methodology to development.

Obviously I'm frustrated. Twill CMS has some very nice features, but this approach to features I am very positive longterm will see me migrate to other systems instead. It's just too frustrating.

**Comment by ifox** [comment]
What are you even talking about? Modules are not going anywhere.

**Comment by ifox** [comment]
Nobody said anywhere that capsules are the way to build Twill modules. Look a little bit further before ranting like this...

**Reply by kirkbushell** [reply]
With all due respect, I wouldn't be ranting if the update didn't force me to register capsules. I installed, ran my tests, broken tests everywhere: No capsule found for: Article - seems to be related to the HasSlug trait.

**Comment by ifox** [comment]
Capsules are simply a way to structure a Twill module. You don't have to use it if you don't like it, like, ever, and we are not pushing people to use it either. As you said yourself, it's not even properly documented. The docs are still talking about modules all over the place and that's not changing. We refactored capsules so they could be autoloaded from our new [package](https://twill.io/docs/packages/) feature, that's it.

>Sigh, why weren't capsules kept for version 3, as it's a HUGE breaking change!?!

You're overreacting so much that your title doesn't even match what you are complaining about...

**Reply by kirkbushell** [reply]
I disagree, the software has forced me to register capsules (see response above), that's why I was wondering why the hell it's not a major release push rather than a minor, despite the claim that it doesn't require it.

**Reply by ifox** [reply]
Your tone of voice prevented me from understanding that. Try to calm down before posting next time, we're not deserving any of it.

Capsules have been part of Twill since 2.2, this is just a bug that we haven't seen yet.

**Comment by pboivin** [comment]
Hi @kirkbushell, this was an odd read...

I have a good appreciation for your previous comments and thoughts on how to improve the project. In fact, some of your previous comments have been the seed to work on better support for composer packages, moving away from array configuration to more OO ways of extending and composing entities, etc. A lot of that effort is now led by @haringsrob, who's hard at work on refactoring and testing. The road ahead is long but we indeed have left the station.

It sounds like Twill is simply not what you are looking for in the end. Maybe Twill's take on solving the "CMS" problem is just not the right level of abstraction for you, and that's ok.

**Reply by kirkbushell** [reply]
On the contrary, Twill CMS offers all the features I want, its implementation and these sorts of issues are what get under my skin.

It's good to hear that it's moving away from array configuration, that's a sigh of relief on my end.

**Reply by kirkbushell** [reply]
PS. I've offered multiple times to help out with the implementation, but I understand it's mostly in-house development (and the previous responses I got didn't seem to imply that you'd move away from array config at all).

**Comment by kirkbushell** [comment]
Here are some details regarding the issues. This was right after upgrade to latest:

    A17\Twill\Exceptions\NoCapsuleFoundException: Article

Stack trace:

/Users/kirkbushell/Work/Mine/rathetimes/vendor/area17/twill/src/TwillCapsules.php:70
/Users/kirkbushell/Work/Mine/rathetimes/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php:261
/Users/kirkbushell/Work/Mine/rathetimes/vendor/area17/twill/src/Models/Behaviors/HasSlug.php:60
/Users/kirkbushell/Work/Mine/rathetimes/vendor/area17/twill/src/Models/Behaviors/HasSlug.php:34
/Users/kirkbushell/Work/Mine/rathetimes/vendor/area17/twill/src/Models/Behaviors/HasSlug.php:155
/Users/kirkbushell/Work/Mine/rathetimes/vendor/area17/twill/src/Models/Behaviors/HasSlug.php:135
/Users/kirkbushell/Work/Mine/rathetimes/vendor/area17/twill/src/Models/Behaviors/HasSlug.php:116
/Users/kirkbushell/Work/Mine/rathetimes/vendor/area17/twill/src/Models/Behaviors/HasSlug.php:15
/Users/kirkbushell/Work/Mine/rathetimes/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php:404
/Users/kirkbushell/Work/Mine/rathetimes/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php:249
/Users/kirkbushell/Work/Mine/rathetimes/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/HasEvents.php:189
/Users/kirkbushell/Work/Mine/rathetimes/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php:1174
/Users/kirkbushell/Work/Mine/rathetimes/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php:994
/Users/kirkbushell/Work/Mine/rathetimes/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:314
/Users/kirkbushell/Work/Mine/rathetimes/vendor/laravel/framework/src/Illuminate/Collections/Traits/EnumeratesValues.php:244
/Users/kirkbushell/Work/Mine/rathetimes/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:317
/Users/kirkbushell/Work/Mine/rathetimes/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:261
/Users/kirkbushell/Work/Mine/rathetimes/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:255
/Users/kirkbushell/Work/Mine/rathetimes/tests/Feature/ThreadTest.php:27

TwillCapsules.php:

```php
public function getCapsuleForModel(string $model): Capsule
    {
        $capsule = $this->getRegisteredCapsules()->first(function (Capsule $capsule) use ($model) {
            return $capsule->getSingular() === $model;
        });

        if (! $capsule) {
            throw new NoCapsuleFoundException($model);
        }

        return $capsule;
    }
```

This happened while trying to utilise HasSlug at version: 2.8.2

Now admittedly I wrote this when I was pissed off and should have taken the night, but even so - this isn't the first time it's happened, that's why I started it as a discussion piece. The claims that capsules aren't required appears to be completely false on at -least- this version, perhaps earlier.

**Reply by ifox** [reply]
You found a bug. That's all it is. Next time submit an issue and be respectful of the work our community is doing instead of claiming we don't know how to release software. We're yet to see a PR from you.

**Reply by ifox** [reply]
https://github.com/area17/twill/blob/b32623b59c5bee4094f072419fed33067698cbe6/src/Models/Behaviors/HasSlug.php#L54

It's not finding your slug model and now we also search if you have a capsule for it. So you probably already had an issue by not having a slug class on the proper namespace while using the HasSlug trait, and we just need to improve the error message.

**Reply by ifox** [reply]
To be clear, no, you absolutely do not need to register capsules and that bug was not intentional.

**Reply by kirkbushell** [reply]
> To be clear, no, you absolutely do not need to register capsules and that bug was not intentional.

Fine, but I had to to get it to work. So the bug results in you actually having to setup capsules.

**Reply by kirkbushell** [reply]
> https://github.com/area17/twill/blob/b32623b59c5bee4094f072419fed33067698cbe6/src/Models/Behaviors/HasSlug.php#L54
> 
> It's not finding your slug model and now we also search if you have a capsule for it. So you probably already had an issue by not having a slug class on the proper namespace while using the HasSlug trait, and we just need to improve the error message.

It was never an issue prior to updating. Been using slugs since day 1.

**Reply by kirkbushell** [reply]
If anything, this bug highlights just how much of an issue it is, if I'm having to setup capsules and reverse-engineer code. I'm still not exactly sure what the bug is or how to fix it, but it's expecting a capsule to be setup for any Model that uses it, it seems.

**Reply by antonioribeiro** [reply]
> Fine, but I had to to get it to work. So the bug results in you actually having to setup capsules.

## The bug

Is a result of:

1. Your model using HasSlug
2. The HasSlug trait was not able to find the Slug Model class related to this Twill module, again, "Twill module" **NOT** "Twill Capsule".

Meaning you have a bug on your code that can be, possibly, fixed either by:

1. Removing the HasSlug from your model
2. Creating or fixing the name the Slug model file

## The error message...

...that you get about it not finding the Capsule is not a bug, it's just a misleading message, because the HasSlug trait, when not finding your Slug class is assuming that it's not your fault, so it tries to get the Slug model from a possible existing Capsule, that obviously doesn't exists.

## It is your fault, Kirk, but that's okay

It's hard to tell without seeing some code, but I believe you know how debugging works, right? Throwing some guesses here: you may have deleted the Slug model, you may have renamed the Twill module and forgot to rename the Slug model, or when you created the module you didn't tell Twill you would be using slugs, so the Slug module was not created for you, and then you for some reason we also don't know, you added the HasSlug to your model, or whatever..., and here we are.

## What we can do to improve

Now that we know that this can be a problem for some users, we can probably add a check on the `getCapsuleForModel()` method to see if there's a Capsule enabled for it and if there is one, check if we can get the Slug model from it, and try to not mislead any other user to think they need to have Capsules enabled for Twill to work.

## What you can do to improve

Next time just open an issue describing clearly what your problem is, like everyone else, or hit us on Discord. I'm unfortunately not as active as I would like to be there, but @ifox (and @haringsrob) are pretty much living a life on that thing, answering every single person that hit us there, from the most elementary questions up to the challenging ones.

## About Capsules, again

As you were told so many times, in so many different ways and conversation topics, that Capsules are completely optional, so let's not repeat ourselves again in the future, this is a huge waste of time. Just forget about them if you don't like them, please. But don't keep trying to make us get rid of Capsules because you think you know better. Capsules are here since [Twill 2.2](https://github.com/area17/twill/releases/tag/2.2.0), this is Twill 2.8, and no one experienced any major troubles with Capsules, everyone that needed help setting them up was able to reach out to us on Discord or here to fix it.

## About Open Source Software

You know exactly what OSS is and the roles that everyone should be playing on it, I will not patronize you.

**Reply by kirkbushell** [reply]
Incorrect, I've been using HasSlug since the beginning. It was working fine until I updated.

**Comment by kirkbushell** [comment]
Let me add as I've added it as comments earlier: I apologise for my hasty comment. I wasn't in a good mood after the upgrade and certainly should have cooled down first before posting - I'm sorry I reacted in such a way.

**Comment by haringsrob** [comment]
Hey @kirkbushell,

Trying to reproduce the issue. Could you check the existence of:

`App\Models\Slugs\ArticleSlug` and `App\Models\Article`

Also please make sure that the namespace and filenames are correct inside the files.

The error you experience should only show up if the `App\Models\Slugs\ArticleSlug` class is not existing or has a bad name.

Since you are upgrading from an older version of twill, it might have to do with the `config('twill.namespace')` setting. Do you have this set to a custom config? That could maybe explain.

You could also add `dd($slug)` to see what class it is trying to load in order to debug this issue:

in `vendor/area17/twill/src/Models/Behaviors/HasSlug.php`

```php
    public function getSlugModelClass()
    {
        $slug = $this->getNamespace() . "\Slugs\\" . $this->getSlugClassName();
        
        dd($slug); // ADD THIS LINE FOR DEBUGGING.

        if (@class_exists($slug)) {
            return $slug;
        }

        return TwillCapsules::getCapsuleForModel(class_basename($this))->getSlugModel();
    }
```

Thanks

**Comment by haringsrob** [comment]
There was another case which provided more information on the issue, mostly it was related to models being in nested folders so:

app/models/somefolder/model instead of /app/models/model

This pr fixes this https://github.com/area17/twill/pull/1738 (at the time of writing not merged).
